lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -209,26 +300,267 @@ classifyCodingKeys(const DerivedConformance &derived) {
+///
+/// This is only useful once a \c CodingKeys enum has been validated (via \c
+/// hasValidCodingKeysEnum) or synthesized (via \c synthesizeCodingKeysEnum).
+///
+/// \param C The \c ASTContext to perform the lookup in.
+///
+/// \param target The target type to look in.
+///
+/// \return A retrieved canonical \c CodingKeys enum if \c target has a valid
+/// one; \c nullptr otherwise.
+static EnumDecl *lookupEvaluatedCodingKeysEnum(ASTContext &C,
+                                               NominalTypeDecl *target) {
+  auto codingKeyDecls = target->lookupDirect(DeclName(C.Id_CodingKeys));
+  if (codingKeyDecls.empty())
+    return nullptr;
+
+  auto *codingKeysDecl = codingKeyDecls.front();
+  if (auto *typealiasDecl = dyn_cast<TypeAliasDecl>(codingKeysDecl))
+    codingKeysDecl = typealiasDecl->getDeclaredInterfaceType()->getAnyNominal();
+
+  return dyn_cast<EnumDecl>(codingKeysDecl);
+}
+
+static EnumDecl *lookupEvaluatedCodingKeysEnum(ASTContext &C,
+                                               NominalTypeDecl *target,
+                                               Identifier identifier) {
+  auto codingKeyDecls = target->lookupDirect(DeclName(identifier));
+  if (codingKeyDecls.empty())
+    return nullptr;
+
+  auto *codingKeysDecl = codingKeyDecls.front();
+  if (auto *typealiasDecl = dyn_cast<TypeAliasDecl>(codingKeysDecl))
+    codingKeysDecl = typealiasDecl->getDeclaredInterfaceType()->getAnyNominal();
+
+  return dyn_cast<EnumDecl>(codingKeysDecl);
+}
+
+static EnumElementDecl *lookupEnumCase(ASTContext &C, NominalTypeDecl *target,
+                                       Identifier identifier) {
+  auto elementDecls = target->lookupDirect(DeclName(identifier));
+  if (elementDecls.empty())
+    return nullptr;
+
+  auto *elementDecl = elementDecls.front();
+
+  return dyn_cast<EnumElementDecl>(elementDecl);
+}
+
+static NominalTypeDecl *lookupErrorContext(ASTContext &C,
+                                           NominalTypeDecl *errorDecl) {
+  auto elementDecls = errorDecl->lookupDirect(C.Id_Context);
+  if (elementDecls.empty())
+    return nullptr;
+
+  auto *decl = elementDecls.front();
+
+  return dyn_cast<NominalTypeDecl>(decl);
+}
+
+/// Synthesizes a new \c CodingKeys enum based on the {En,De}codable members of
+/// the given type (\c nullptr if unable to synthesize).
+///
+/// If able to synthesize the enum, adds it directly to \c derived.Nominal.
+static bool synthesizeCodingKeysEnum_enum(DerivedConformance &derived,
+                                          EnumDecl *target) {
+  auto &C = derived.Context;
+
+  // We want to look through all the var declarations of this type to create
+  // enum cases based on those var names.
+  auto *codingKeyProto = C.getProtocol(KnownProtocolKind::CodingKey);
+  auto codingKeyType = codingKeyProto->getDeclaredInterfaceType();
+  TypeLoc protoTypeLoc[1] = {TypeLoc::withoutLoc(codingKeyType)};
+  ArrayRef<TypeLoc> inherited = C.AllocateCopy(protoTypeLoc);
```

Comment:
The function only gets called when synthesis is actually necessary. But since we synthesize several `CodingKeys` enums here, some may be defined already. But if this function is called, we definitely need this array.